Remove use of "[]" to access data

ffmancera opened the pull request:
This a good first step on error handling, in a following PR we should raise the error or panic instead of ignore the values.

YtvwlD committed:
nla: rework parse_as_ipv4/6 to return Result<T, NisporError>

They now panic with more descriptive messages when the input is invalid.
Also, they now return a Ipv4Addr or Ipv6Addr and are used
in more places.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

tyll reviewed (commented):
FYI: The commit message does not follow best practices that are documented for Nmstate: https://github.com/nmstate/nmstate/blob/base/CONTRIBUTING.md#write-a-good-commit-message

Especially it is not documenting the "why: ` Explain the problem that this commit is solving. Focus on why you are making 
 this change as opposed to how (the code explains that)."

ffmancera commented:
> FYI: The commit message does not follow best practices that are documented for Nmstate: https://github.com/nmstate/nmstate/blob/base/CONTRIBUTING.md#write-a-good-commit-message
> 
> Especially it is not documenting the "why: ` Explain the problem that this commit is solving. Focus on why you are making
> this change as opposed to how (the code explains that)."

Thank you @tyll! Yes, I will reword the commit title and messages. Thank you 👍

cathay4t reviewed (changes requested):
This PR is not improving but make the code worse when debugging problems.

Instead, I would like to see we change functions return `Result<> `and  `?` for converting
it to `NisporError`.

Meanwhile, please don't use `.ok()` which hide the error message.

ffmancera commented:
> This PR is not improving but make the code worse when debugging problems.
> 
> Instead, I would like to see we change functions return `Result<> `and `?` for converting
> it to `NisporError`.
> 
> Meanwhile, please don't use `.ok()` which hide the error message.

In that case, I will modify all the functions to return Result<>. Thanks.

ffmancera committed:
all: use "get()" instead of "[]"

This patch is replacing all the "[]" for "get()". The functions are now
returning Result<T, NisporError> which will make debugging easier and
the users will be able to handle the exceptions properly instead of
panicking.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

ffmancera commented:
> This PR is not improving but make the code worse when debugging problems.
> 
> Instead, I would like to see we change functions return `Result<> `and `?` for converting
> it to `NisporError`.
> 
> Meanwhile, please don't use `.ok()` which hide the error message.

Done :-)

cathay4t commented on src/lib/ifaces/iface.rs:
```diff
@@ -279,9 +284,11 @@ pub(crate) fn parse_nl_msg_to_iface(nl_msg: &LinkMessage) -> Option<Iface> {
                     // Remove the tailing \0
-                    match std::str::from_utf8(&(d.as_slice()[0..(d.len() - 1)]))
+                    match std::ffi::CStr::from_bytes_with_nul(&d.as_slice())
+                        .ok()
```
Please replace the use of `.ok().and_then()` to `?`

ffmancera replied:
Done, thank you!

cathay4t commented on src/lib/error.rs:
```diff
@@ -57,3 +57,12 @@ impl std::convert::From<std::io::Error> for NisporError {
     }
 }
+
+impl std::convert::From<&str> for NisporError {
```
Converting `&str` to `NisporError` seems too dangerous to me, normally we only convert other type of errors into NisporError.

You may try to implement the `NisporError::bug(msg: &str)` to simply the effort on create a `NisporError` with `ErrorKind::NisporBug`

ffmancera replied:
Done! I've implemented the bug() function for NisporError. Thanks!

cathay4t commented on src/lib/error.rs:
```diff
@@ -40,6 +49,24 @@ impl std::convert::From<rtnetlink::Error> for NisporError {
+    fn from(e: std::ffi::FromBytesWithNulError) -> Self {
+        NisporError {
+            kind: ErrorKind::NisporBug,
+            msg: e.to_string(),
```
How about `format!("FromBytesWithNulError: {}", e)` so that we could know the source this error.

ffmancera replied:
Done! Thanks!

cathay4t commented on src/lib/error.rs:
```diff
@@ -40,6 +49,24 @@ impl std::convert::From<rtnetlink::Error> for NisporError {
+    fn from(e: std::str::Utf8Error) -> Self {
+        NisporError {
+            kind: ErrorKind::NisporBug,
+            msg: e.to_string(),
```
How about `format!("Utf8Error: {}", e)` so that we could know the source this error.

ffmancera replied:
Done, thanks!
